### ISSUE TITLE
Regression in v0.19 when using relative imports and paths from tsconfig

### ISSUE THREAD

**Issue by andyvanee** [evanw/esbuild#3354]
I believe this may be related to other issues, but haven't seen a minimal reproduction expressed elsewhere:

[Expected output - v0.18.20 processes relative import correctly](https://esbuild.github.io/try/#YgAwLjE4LjIwAC0tYnVuZGxlIC0tZm9ybWF0PWVzbSAtLXBsYXRmb3JtPW5vZGUgLS1vdXRkaXI9ZGlzdABlAGVudHJ5LnRzAGltcG9ydCB7Zm9vfSBmcm9tICJmb28iOwppbXBvcnQgcHJvY2VzcyBmcm9tICJub2RlOnByb2Nlc3MiOwoKZm9vKCk7AAB0c2NvbmZpZy5qc29uAHsKICAiY29tcGlsZXJPcHRpb25zIjogewogICAgImJhc2VVcmwiOiAiLiIsCiAgICAicGF0aHMiOiB7CiAgICAgICIqIjogWyJ3ZWIvKiJdCiAgICB9CiAgfQp9AAB3ZWIvZm9vLnRzAGltcG9ydCB7Zm9vIGFzIGJhckZvb30gZnJvbSAnYmFyL2Zvbyc7CgpleHBvcnQgZnVuY3Rpb24gZm9vKCkgewogIGNvbnNvbGUubG9nKCd3ZWIvZm9vJyk7CiAgYmFyRm9vKCk7Cn0AAHdlYi9iYXIvZm9vL2Zvby50cwBleHBvcnQgZnVuY3Rpb24gZm9vKCkgewogIGNvbnNvbGUubG9nKCdiYXIvZm9vJyk7Cn07AAB3ZWIvYmFyL2Zvby9pbmRleC50cwAvLyBUaGlzIHJlbGF0aXZlIGltcG9ydCBpcyBwcm9ibGVtYXRpYyBpbiB2MC4xOQovLyBzaW5jZSBpdCByZXNvbHZlcyB0byBgd2ViL2Zvb2AgcmF0aGVyIHRoYW4gYHdlYi9iYXIvZm9vYApleHBvcnQge2Zvb30gZnJvbSAnLi9mb28n)

[Regression - v0.19 resolves the relative import incorrectly](https://esbuild.github.io/try/#YgAwLjE5LjIALS1idW5kbGUgLS1mb3JtYXQ9ZXNtIC0tcGxhdGZvcm09bm9kZSAtLW91dGRpcj1kaXN0AGUAZW50cnkudHMAaW1wb3J0IHtmb299IGZyb20gImZvbyI7CmltcG9ydCBwcm9jZXNzIGZyb20gIm5vZGU6cHJvY2VzcyI7Cgpmb28oKTsAAHRzY29uZmlnLmpzb24AewogICJjb21waWxlck9wdGlvbnMiOiB7CiAgICAiYmFzZVVybCI6ICIuIiwKICAgICJwYXRocyI6IHsKICAgICAgIioiOiBbIndlYi8qIl0KICAgIH0KICB9Cn0AAHdlYi9mb28udHMAaW1wb3J0IHtmb28gYXMgYmFyRm9vfSBmcm9tICdiYXIvZm9vJzsKCmV4cG9ydCBmdW5jdGlvbiBmb28oKSB7CiAgY29uc29sZS5sb2coJ3dlYi9mb28nKTsKICBiYXJGb28oKTsKfQAAd2ViL2Jhci9mb28vZm9vLnRzAGV4cG9ydCBmdW5jdGlvbiBmb28oKSB7CiAgY29uc29sZS5sb2coJ2Jhci9mb28nKTsKfTsAAHdlYi9iYXIvZm9vL2luZGV4LnRzAC8vIFRoaXMgcmVsYXRpdmUgaW1wb3J0IGlzIHByb2JsZW1hdGljIGluIHYwLjE5Ci8vIHNpbmNlIGl0IHJlc29sdmVzIHRvIGB3ZWIvZm9vYCByYXRoZXIgdGhhbiBgd2ViL2Jhci9mb29gCmV4cG9ydCB7Zm9vfSBmcm9tICcuL2Zvbyc)

The most relevant pieces to this reproduction are:

1. Including a wildcard in `compilerOptions.paths`
2. Using a relative import which *could be* resolved by a path-relative import

```json
{
  "compilerOptions": {
    "paths": {
      "*": ["web/*"]
    }
}
```

```ts
// web/bar/foo/index.ts
// This relative import is problematic in v0.19 since it resolves to `web/foo` rather than `web/bar/foo`
export {foo} from './foo'
```

The output in `v0.19` translates the example into a recursive function since `web/bar/foo/index.ts` is masked by `web/foo.ts`

```js
// web/foo.ts
function foo() {
  console.log("web/foo");
  foo();                                  // <- this is the wrong function reference
}

// entry.ts
foo();
```

- I believe this regression was introduced in the resolution to https://github.com/evanw/esbuild/issues/3238

**Comment by andyvanee** [comment]
I have resolved this issue by adding an explicit path aliases for relative imports, but this seems like a pretty strange workaround:

```json
{
  "compilerOptions": {
    "baseUrl": ".",
    "paths": {
      ".": ["."],
      "./*": ["./*"],
      "*": ["web/*"]
    }
  }
}
``

**Comment by taylorthurlow** [comment]
Same issue, and @andyvanee's workaround also worked for me - thanks for posting this.

**Comment by briancavalier** [comment]
Seeing a similar issue. Doesn't seem to be exactly the same context, but may have the same root cause.  For us, it's manifesting as our build missing an export from package in the same monorepo as an application.  Reverting to 0.18.20 works for us.  We haven't tried @andyvanee's workaround of adding explicit relative import paths yet.  Will report back if we try that.

**Comment by ilijapuaca** [comment]
We seem to have a similar issues, import paths (non-relative) which are set up as aliases through `paths` no longer get resolved properly since `0.19.0`. We are not marking any of these packages as external, and the workaround above did not seem to work

**Comment by steabert** [comment]
> We seem to have a similar issues, import paths (non-relative) which are set up as aliases through `paths` no longer get resolved properly since `0.19.0`. We are not marking any of these packages as external, and the workaround above did not seem to work

There was an update with a fix for the resolution behaviour in v0.19.1, you can retry the work-around with that version.

**Comment by ilijapuaca** [comment]
> There was an update with a fix for the resolution behaviour in v0.19.1, you can retry the work-around with that version.

I tried both 0.19.1 and 0.19.2, having the same issue

**Comment by briancavalier** [comment]
> I tried both 0.19.1 and 0.19.2, having the same issue

Same here.  0.19.0-0.19.2 reproduce our issue.  We haven't tried 0.19.3 yet.

**Comment by andyvanee** [comment]
> There was an update with a fix for the resolution behaviour in v0.19.1, you can retry the work-around with that version.

Without seeing a reproduction case, I'm not sure if @ilijapuaca is having the same issue, but you can see from my original reproduction that none of the `0.19.x` versions address the issue: [Regression - v0.19 resolves the relative import incorrectly](https://esbuild.github.io/try/#YgAwLjE5LjIALS1idW5kbGUgLS1mb3JtYXQ9ZXNtIC0tcGxhdGZvcm09bm9kZSAtLW91dGRpcj1kaXN0AGUAZW50cnkudHMAaW1wb3J0IHtmb299IGZyb20gImZvbyI7CmltcG9ydCBwcm9jZXNzIGZyb20gIm5vZGU6cHJvY2VzcyI7Cgpmb28oKTsAAHRzY29uZmlnLmpzb24AewogICJjb21waWxlck9wdGlvbnMiOiB7CiAgICAiYmFzZVVybCI6ICIuIiwKICAgICJwYXRocyI6IHsKICAgICAgIioiOiBbIndlYi8qIl0KICAgIH0KICB9Cn0AAHdlYi9mb28udHMAaW1wb3J0IHtmb28gYXMgYmFyRm9vfSBmcm9tICdiYXIvZm9vJzsKCmV4cG9ydCBmdW5jdGlvbiBmb28oKSB7CiAgY29uc29sZS5sb2coJ3dlYi9mb28nKTsKICBiYXJGb28oKTsKfQAAd2ViL2Jhci9mb28vZm9vLnRzAGV4cG9ydCBmdW5jdGlvbiBmb28oKSB7CiAgY29uc29sZS5sb2coJ2Jhci9mb28nKTsKfTsAAHdlYi9iYXIvZm9vL2luZGV4LnRzAC8vIFRoaXMgcmVsYXRpdmUgaW1wb3J0IGlzIHByb2JsZW1hdGljIGluIHYwLjE5Ci8vIHNpbmNlIGl0IHJlc29sdmVzIHRvIGB3ZWIvZm9vYCByYXRoZXIgdGhhbiBgd2ViL2Jhci9mb29gCmV4cG9ydCB7Zm9vfSBmcm9tICcuL2Zvbyc)

The expected output should be two function definitions as it is in `0.18.x`:

```javascript
// web/bar/foo/foo.ts
function foo() {
  console.log("bar/foo");
}

// web/foo.ts
function foo2() {
  console.log("web/foo");
  foo();
}

// entry.ts
foo2();
```

While the actual output from `0.19.x` is an infinitely recursive function:

```javascript
// web/foo.ts
function foo() {
  console.log("web/foo");
  foo();
}

// entry.ts
foo();
```

**Comment by steabert** [comment]
> > There was an update with a fix for the resolution behaviour in v0.19.1, you can retry the work-around with that version.
> 
> I tried both 0.19.1 and 0.19.2, having the same issue

Note that I was only referring to the _work-around_ needing 0.19.1 (for me at least). The issue still remains.

**Comment by lnist** [comment]
Possibly related, we had to change `import * as typechain from "./typechain";` to `import * as typechain from "./typechain/index";` otherwise esbuild picks up the node_modules/typechain path instead of the local "./typechain".
